### ISSUE TITLE
Secure file transfer

### ISSUE THREAD

**Issue by fakuivan** [localsend/localsend#162]
From what I understand HTTPS transfers are secured using a self signed cert. Is there any way to check if the same cert is being used by both parties? If the reciever or sender accept any random cert it seems like this could be open to a MITM attack.

**Comment by Tienisto** [comment]
I am not a security expert but before you accept the request, you can verify if the alias and hashtag (IP) match. Especially the IP address is difficult to modify (afaik).

I can add an option to allow users to compare their certs though, even if this will be rarely used

**Comment by fakuivan** [comment]
The certificate fingerprint could be put into a QR code to then be checked using the camera (similarly to what whatsapp does with contacts). If a device persists its certificate used for transfers to it, the QR code could be scanned only the first time. There are other more involved options like CAs (certificate autorities) if you want to still use temporary certs for each transfer.

**Comment by hardcore-sushi** [comment]
> before you accept the request, you can verify if the alias and hashtag (IP) match. Especially the IP address is difficult to modify (afaik).

Unfortunately, in local area networks, the connection can be easily intercepted using ARP spoofing or other techniques. If the connection is not authenticated, then the attacker can replace the users certificates with their own and decrypt or modify all the exchanged traffic, without any users noticing it. This is why TLS is usually used with a [PKI](https://en.wikipedia.org/wiki/Public_key_infrastructure).

From what I understand from [the protocol documentation](https://github.com/localsend/protocol#2-fingerprint), certificates are already hashed and provided as a fingerprint to the others devices on the network. If this fingerprint was displayed to the users, or even better, used to deterministically derive the device alias, the app would only have to verify at the connection time that both the HTTPS server and the client provide certificates that actually match the advertised fingerprints. That way, users would only need to compare the fingerprints displayed on both device (or derived alias) to ensure the connection is secure (no risk of MITM attack).

To improve UX, you could allow users to remember devices along with their fingerprint and mark them as trusted. Thus, future connections to saved trusted devices would not need to be authenticated again.

**Comment by 2shrestha22** [comment]
I am wondering how HTTPS is secure when the server is a client side app and attacker may easily get certificate the server is using.

**Comment by 2shrestha22** [comment]
> I am wondering how HTTPS is secure when the server is a client side app and attacker may easily get certificate the server is using.

Ok so the certificates are generated on app start. And the client does not need server certificate?

**Comment by wuwbobo2021** [comment]
Excuse me, I am new to this application, and I know little about network security concepts.

However, [CVE-2025-54792](https://github.com/localsend/localsend/security/advisories/GHSA-424h-5f6m-x63f) really makes me confused. Many websites claims that this vulnerability is fixed in the current version (1.17.0) of `localsend`, but some of them also marked this current version as vulnerable, see <https://www.cvedetails.com/cve/CVE-2025-54792>. According to these sites' notes, There is a "patch" for this vulnerability, but this patch actually fixes [CVE‑2025‑27142](https://github.com/localsend/localsend/security/advisories/GHSA-f7jp-p6j4-3522). In my understanding, this is just lowering the severity of CVE-2025-54792, which is not actually being fixed at this moment. This confusion looks a bit crazy to me...

The `Remediation` section of CVE-2025-54792 also makes me confused:

> Implement TCP handshake for IP address verification. The "Favorites" heart icon *sould* only follow the right IP instead of anything else.

What is the concept of IP address verification? If the attacker is connected to the target users' wireless router, at least it may set the IP address of itself to an IP address that had been used by one of the users (who could be offline for a moment).

**Comment by fakuivan** [comment]
@wuwbobo2021 if they're not checking certs then encryption and authentication is easily bypassed using arp spoofing. There is really no solution besides checking certs. I stopped using the app once I noticed this vulnerability, so I'm not aware of how modern versions work.

**Comment by JoeyC-Dev** [comment]
> you can verify if the alias and hashtag (IP) match. Especially the IP address is difficult to modify (afaik).

@Tienisto I want to understand whether this assumption still holds in LocalSend today: that the peer IP address is difficult to spoof, manipulate or change.

I thought that CVE report already made it clear.

**Comment by fakuivan** [comment]
Current issue seems to still be present and tracked as CVE-2025-54792 / GHSA-424h-5f6m-x63f

More info: https://chatgpt.com/share/6a259eb2-4734-83e9-af97-4e34623c1cbe

**Comment by JoeyC-Dev** [comment]
> Current issue seems to still be present and tracked as CVE-2025-54792 / [GHSA-424h-5f6m-x63f](https://github.com/localsend/localsend/security/advisories/GHSA-424h-5f6m-x63f)

I don’t think this issue is going to be fixed anytime soon, given that it has already been open for 10 months. And if seriously tracking, you already mentioned it since 2023. 

I’d like to understand the maintainers’ plan for this issue. So far, I haven’t seen any progress updates or announcements (I already checked commit history, there is none for this CVE fix).

BTW, I may release a PoC soon. Demo: https://mastodon.social/@joeycdev/116689754693618990

**Comment by JoeyC-Dev** [comment]
Write-up/PoC: https://blog.joeyc.dev/posts/localsend-cve-2025-54792/

Vibed patch: https://github.com/JoeyC-Dev/localsend-cve-2025-54792-vibed-patch

**Disclaimer:** I am not Flutter developer, so I had to use 3 days to "vibe" the proposed fix. I have tested it on Windows and Android (with compatible with v1.17.0). Use it at your own risk.

~~Sigh. Open-source.~~

---
> I am not a security expert 

I will take this lesson, from this 83.1k star project.

---

> Current issue seems to still be present and tracked as CVE-2025-54792 / [GHSA-424h-5f6m-x63f](https://github.com/localsend/localsend/security/advisories/GHSA-424h-5f6m-x63f)
> 
> More info: https://chatgpt.com/share/6a259eb2-4734-83e9-af97-4e34623c1cbe

BTW, if it is not Pro thinking, GPT will return below result:

> the issue was fixed in 1.17.0.

<img width="809" height="424" alt="Image" src="https://github.com/user-attachments/assets/ff42300c-0426-40db-9274-d23ddba13bcf" />

**Comment by hollunder** [comment]
Maybe I misunderstand something, but if localsend already use a self-signed certificate, couldn't it simply pin that very certificate?

**Comment by fakuivan** [comment]
@hollunder Even if you do pin it, which is good, you need a way to compare them so you pin the correct one.

**Comment by Tienisto** [comment]
Hi, the Rust rewrite of the networking layer is complete. The cert should be pinned properly in main and in the next release v1.18.0.

**Comment by JoeyC-Dev** [comment]
> Hi, the Rust rewrite of the networking layer is complete. The cert should be pinned properly in main and in the next release v1.18.0.

Analysis copied from GPT-5.6 Sol, for reference only:

<details>

Verdict: **partially fixed, but not fully**.

The author’s branch correctly fixes the most dangerous send-side consequence: when a user selects an HTTPS device, the sender pins the connection to that device’s advertised TLS certificate fingerprint before `prepare-upload`, file upload, or cancel requests. A mismatched server certificate fails during the TLS handshake, before metadata or file bytes are sent. This is implemented in [send_provider.dart](app/lib/provider/network/send_provider.dart:67), [http_provider.dart](app/lib/provider/http_provider.dart:24), and [server_cert_verifier.rs](packages/core/src/http/client/server_cert_verifier.rs:56). The upstream tests explicitly verify that an impostor receives neither upload metadata nor file content.

It also adds mTLS-side validation for incoming `/register`: the server only emits a registration event when the JSON `fingerprint` equals the SHA-256 of the sender’s TLS client certificate. See [v2.rs](packages/core/src/http/server/v2.rs:150).

However, it **does not fully fix multicast discovery poisoning**:

```text
UDP multicast payload
  -> converted directly to Device
  -> yielded to app
  -> RegisterDeviceAction stores it in nearby-device state
  -> only afterward, an unpinned /register response is attempted
```

The raw UDP device is yielded at [multicast_discovery.dart](packages/localsend_isolates/lib/src/task/discovery/multicast_discovery.dart:70), and the app immediately stores it at [nearby_devices_provider.dart](app/lib/provider/network/nearby_devices_provider.dart:52). No certificate check occurs before that registration.

So an attacker can still:

- inject arbitrary fake devices into the nearby list;
- claim another device’s fingerprint in a UDP packet;
- alter a favorite’s automatically updated alias, because [RegisterDeviceAction](app/lib/provider/network/nearby_devices_provider.dart:84) trusts the unverified device fingerprint;
- cause an automatic `/register` response to an endpoint chosen by the multicast packet. That response uses the unpinned discovery client at [multicast_discovery.dart](packages/localsend_isolates/lib/src/task/discovery/multicast_discovery.dart:109), leaking basic local device metadata to that endpoint.

The new pinning means a fake entry claiming a real fingerprint should not successfully receive files, because it cannot present the matching certificate. That is a substantial mitigation of the CVE’s transfer/impersonation impact.

But if the CVE’s fix requirement is “do not trust spoofable UDP discovery data to populate or modify the device cache,” then **no, this branch has not fully fixed it**. The remaining correct design is to treat UDP only as a hint, then verify `TLS certificate hash == UDP-claimed fingerprint` before registering or updating that device.
</details>

I see author is already using Claude. Maybe it helps on fixing CVE.
